test/jdk/java/util/stream/GatherersMapConcurrentTest.java

```diff
@@ -0,0 +1,277 @@
+ * questions.
+ */
+
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Semaphore;
+import java.util.stream.Gatherer;
+import java.util.stream.Gatherers;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assumptions.*;
+
+/**
+ * @test
+ * @summary Tests the API and contract of Gatherers.mapConcurrent
+ * @enablePreview
+ * @run junit GatherersMapConcurrentTest
+ */
+
+public class GatherersMapConcurrentTest {
+
+    record Config(int streamSize, boolean parallel) {
+        Stream<Integer> stream() {
+            var stream = Stream.iterate(1, i -> i + 1).limit(streamSize);
+            stream = parallel ? stream.parallel() : stream.sequential();
+            return stream;
+        }
+    }
+
+    static final Stream<Integer> sizes(){
+        return Stream.of(0,1,10,33,99,9999);
+    }
+
+    static final Stream<Config> sequentialAndParallel(int size) {
+        return Stream.of(false, true)
+                .map(parallel ->
+                        new Config(size, parallel));
+    }
+
+    static final Stream<Config> configurations() {
+        return sizes().flatMap(i -> sequentialAndParallel(i));
+    }
+
+    static final Stream<Config> small_atleast3_configurations() {
+        return sizes().filter(i -> i > 2 && i < 100).flatMap(i -> sequentialAndParallel(i));
+    }
+
+    static final class TestException extends RuntimeException {
+        TestException(String message) {
+            super(message);
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = { Integer.MIN_VALUE, -999, -1, 0})
+    public void throwsIAEWhenConcurrencyLevelIsLowerThanOne(int level) {
+        assertThrows(IllegalArgumentException.class,
+                () -> Gatherers.<String, String>mapConcurrent(level, s -> s));
+    }
+
+    @Test
+    public void throwsNPEWhenMapperFunctionIsNull() {
+        assertThrows(NullPointerException.class, () -> Gatherers.<String, String>mapConcurrent(2, null));
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false } )
+    public void rethrowsRuntimeExceptionsUnwrapped(boolean parallel) {
+        final var stream = parallel ? Stream.of(1).parallel() : Stream.of(1);
+
+        var exception =
+            assertThrows(
+                RuntimeException.class,
+                     () -> stream.gather(
+                             Gatherers.<Integer, Integer>mapConcurrent(2, x -> {
+                                 throw new RuntimeException("expected");
+                             })
+                           ).toList()
+            );
+        assertEquals("expected", exception.getMessage());
+        assertNull(exception.getCause());
```

Comment:
I assume it would be possible to add a comrade to this test for the wrapping case too, e.g. if the mapper throws an Error.

Reply:
@AlanBateman Yes, good point. I'll add that as well now :+1: